### ISSUE TITLE
More awesome HyperApp

### ISSUE THREAD

**Issue by infinnie** [infinnie/infinnie.github.io#11]
```jsx
define.require(["jquery", "vendors/app", "vendors/h", "pages/login", "vendors/md5"], function ($, app, h, LoginPage, md5) {
    "use strict";
    /*@jsx h*/
    $(function () {
        app({
            state: {
                currentPage: null
            },
            view: function (state, actions) {
                return state.currentPage && state.currentPage(state, actions);
            }, root: $("#app").get(0),
            actions: {
                logIn: function (state, actions) {
                    //alert(JSON.stringify(state, null, 4));
                    console.log("Navigating");
                    var d = $.Deferred();
                    setTimeout(function () {
                        d.resolve({
                            company: "Hello world"
                        });
                    }, 233);
                    return $.when(define.requirePromise(["pages/maintainers"]), d.promise()).then(function (arr, result) {
                        var MaintainersPage = arr[0];
                        return {
                            currentPage: function (state, actions) {
                                return <MaintainersPage logOut={actions.logOut} gotoDetails={actions.gotoMaintainerDetails} company={state.pageData.company} />
                            },
                            pageData: { company: result.company }
                        };
                    });
                }, inputToggle: function (state, actions, update) {
                    var ret = {};
                    ret[update.control] = $.extend({}, state.pageData[update.control], { focus: update.value });
                    return {
                        pageData: $.extend({}, state.pageData, ret)
                    };
                }, inputChange: function (state, actions, update) {
                    var ret = {};
                    ret[update.control] = $.extend({}, state.pageData[update.control], { value: update.value });
                    return {
                        pageData: $.extend({}, state.pageData, ret)
                    };
                }, logOut: function () {
                    var d = $.Deferred();
                    setTimeout(function () {
                        d.resolve({
                            currentPage: function (state, actions) {
                                return <LoginPage submit={function () {
                                                    actions.logIn && actions.logIn();
                                                    return false;
                                                }} usernameCtrl={state.pageData.username}
                                                  passwordCtrl={state.pageData.password}
                                                  captchaCtrl={state.pageData.captcha}
                                                  inputToggle={actions.inputToggle}
                                                  inputChange={actions.inputChange
                                } />
                            },
                            pageData: {
                                username: {
                                    ctrl: "username",
                                    focus: false,
                                    value: ""
                                },
                                password: {
                                    ctrl: "password",
                                    focus: false,
                                    value: ""
                                },
                                captcha: {
                                    ctrl: "captcha",
                                    focus: false,
                                    value: ""
                                }
                            }
                        });
                    }, 1);
                    return d.promise();
                }, gotoMaintainerDetails: function () {
                    return define.requirePromise(["pages/maintainerdetails"]).then(function (arr) {
                        var MaintainerDetailsPage = arr[0];
                        return {
                            currentPage: function (state, actions) {
                                return <MaintainerDetailsPage logOut={actions.logOut} gotoList={actions.logIn } />;
                            },
                            pageData: {}
                        };
                    });
                }
            }, events: {
                resolve: function (state, actions, result) {
                    return function (update) {
                        $.when(result).then(update);
                        return result;
                    };
                }, load: function (state, actions, result) {
                    actions.logOut();
                }, render: function () {
                    console.log("Rendering...");
                }
            }
        });
    });
});
```

**Comment by jbucaran** [comment]
@infinnie What is this magic? 😄

**Comment by infinnie** [comment]
Imagine that for every component change (say, in an SPA) one wanted to get both the component (here `define.requirePromise()`) and the initializing data for that component (say, `fetchCompany()`) asynchronously. It would be hard for him or her to ensure that the app renders only once for every component change (I know that it is asynchronous, but…).

Including a function that can return the updated component itself along with the initializing data in the return value of an action, and calling the function in `view()` when it gets rendered, would solve the problem.
